pkg/compactor/blocks_cleaner.go

```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
+			Name: "cortex_bucket_blocks_cleaned_total",
+			Help: "Total number of blocks deleted for a tenant.",
+		}, commonLabels),
+		tenantCleanDuration: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "cortex_bucket_clean_duration_seconds",
+			Help: "Duration of cleaner runtime for a tenant in seconds",
+		}, commonLabels),
 	}
 
-	c.Service = services.NewTimerService(cfg.CleanupInterval, c.starting, c.ticker, nil)
+	c.Service = services.NewBasicService(c.starting, c.loop, nil)
 
 	return c
 }
 
+type cleanerJob struct {
+	users     []string
+	timestamp int64
+}
+
+func (c *BlocksCleaner) SetRingLifecyclerID(ringLifecyclerID string) {
+	c.ringLifecyclerID = ringLifecyclerID
+}
+
 func (c *BlocksCleaner) starting(ctx context.Context) error {
 	// Run a cleanup so that any other service depending on this service
 	// is guaranteed to start once the initial cleanup has been done.
-	c.runCleanup(ctx, true)
+	activeUsers, deletedUsers, err := c.scanUsers(ctx)
 
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to scan users on startup", "err", err.Error())
+		c.runsFailed.WithLabelValues(deletedStatus).Inc()
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
+		return nil
+	}
+	if err = c.cleanUpActiveUsers(ctx, activeUsers, true); err != nil {
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
+	}
+	if err = c.cleanDeletedUsers(ctx, deletedUsers); err != nil {
+		c.runsFailed.WithLabelValues(deletedStatus).Inc()
+	}
 	return nil
 }
 
-func (c *BlocksCleaner) ticker(ctx context.Context) error {
-	c.runCleanup(ctx, false)
+func (c *BlocksCleaner) loop(ctx context.Context) error {
+	t := time.NewTicker(c.cfg.CleanupInterval)
+	defer t.Stop()
 
-	return nil
+	usersChan := make(chan *cleanerJob)
+	deleteChan := make(chan *cleanerJob)
+	defer close(usersChan)
+	defer close(deleteChan)
+
+	go func() {
+		c.runActiveUserCleanup(ctx, usersChan)
+	}()
+	go func() {
+		c.runDeleteUserCleanup(ctx, deleteChan)
+	}()
+
+	for {
+		select {
+		case <-t.C:
+			activeUsers, deletedUsers, err := c.scanUsers(ctx)
+			if err != nil {
+				level.Error(c.logger).Log("msg", "failed to scan users blocks cleanup and maintenance", "err", err.Error())
+				c.runsFailed.WithLabelValues(deletedStatus).Inc()
+				c.runsFailed.WithLabelValues(activeStatus).Inc()
+				continue
+			}
+			cleanJobTimestamp := time.Now().Unix()
+			usersChan <- &cleanerJob{
+				users:     activeUsers,
+				timestamp: cleanJobTimestamp,
+			}
+			deleteChan <- &cleanerJob{
+				users:     deletedUsers,
+				timestamp: cleanJobTimestamp,
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (c *BlocksCleaner) runActiveUserCleanup(ctx context.Context, jobChan chan *cleanerJob) {
+	for job := range jobChan {
+		if job.timestamp < time.Now().Add(-c.cfg.CleanupInterval).Unix() {
+			level.Warn(c.logger).Log("Active user cleaner job too old. Ignoring to get recent data")
+			continue
+		}
+		c.cleanUpActiveUsers(ctx, job.users, false) //nolint:errcheck
```

Comment:
Don't we need to check error here and increase metric if failed? Similar to what we do in `starting`?

Reply:
Addresed